proposals/0083-relax-entry-constraints.md

```diff
@@ -0,0 +1,62 @@
+---
+simd: '0083'
+title: Relax Entry Constraints
+authors:
+  - Andrew Fitzgerald (Solana Labs)
+category: Standard
+type: Core
+status: Draft
+created: 2023-11-02
+feature: (fill in with feature tracking issues once accepted)
+---
+
+## Summary
+
+Remove the constraint that transaction entries must not conflict with each
+other.
+
+## Motivation
+
+The current constraint that transaction entries must not conflict with each
+other is unnecessary.
+It is a constraint, that made the current labs-client implementation easier,
+but is not required for the protocol to function correctly.
+The removal of this constraint simplifies the protocol, and allows more
+flexibility in the implementation of both block-produciton and
+block-validation.
+
+## Alternatives Considered
+
+1. Do nothing
+    - This is the simplest option, as we could leave the protocol as is.
+    However, this leaves the protocol more complex than it needs to be.
+
+## New Terminology
+
+None
+
+## Detailed Design
+
+Currently, if a transaction entry within a block contains transactions that
+conflict, the entire block is marked invalid.
+The proposal is that this constraint is removed entirely, and entries are
+allowed to contain transactions that conflict with each other.
+
+## Impact
+
+- The protocol is simplified
+- Block-production is more flexible
+
+## Security Considerations
+
+None
+
+## Drawbacks
+
+None
+
+## Backwards Compatibility
+
+This proposal is backwards compatible with the current protocol, since it only
+relaxes constraints, and does not add any new constraints.
+All previously valid blocks would still be valid.
```

Comment:
Not to beat a dead horse, but Liam is correct about this being a consensus altering change.

> All previously valid blocks would still be valid.

This is true. However, a block that was previously invalid might now be valid. Or stated another way, a block might be valid under this new set of rules but invalid under the old set. As with other consensus altering changes, this could create a problem during a network upgrade where say 60% is running with old rules / 40% with new rules. As such, this change would need to be rolled out with a feature gate.

Reply:
See other comment here: https://github.com/solana-foundation/solana-improvement-documents/pull/83#discussion_r1381141781

> However, a block that was previously invalid might now be valid.

Your statement is true, but this seems like a **forward* compatibility issue to me.
After the activation of the feature, we don't have previously invalid blocks because they cannot, by definition, be in the history. Therefore all previous (backward) blocks are still valid with the changes

Reply:
>  *forward compatibility issue to me.

Yep, agreed. I guess this gets back to whether a SIMD is implicitly thought of as a consensus breaking change or not. In any case, you're seemingly well aware of the implications of the change so marking this item as resolved, and we can figure out how explicit we need to be in stating consensus implications

Reply:
Added a small comment that newly produced blocks may not be valid under the old protocol: f76e7d3